### PR TITLE
Add arbitrary index field support to FiftyOneTorchDataset

### PR DESCRIPTION
**Pull request by jacobsela**
## Summary

Generalize `FiftyOneTorchDataset` (FOTD) beyond per-sample indexing so a list-valued field path can serve as the dataset's row key.

- New `index_field` parameter on `SampleCollection.to_torch(...)`. Examples:
  - `index_field='frames.id'` — one row per frame doc on video datasets
  - `index_field='ground_truth.detections.id'` — one row per detection
- Broadcast semantics follow the LCA of the index path with each required field: parent fields broadcast across rows sharing an ancestor, child fields vary per row. No flat table of repeated parent values; storage stays normalized.
- Aggregation builds a batched, level-aware pipeline (`_prepare_batch_db`) so multi-level `\$unwind`s aren't paid per row.

## What changed

| File | Notes |
|---|---|
| `fiftyone/utils/torch.py` | Core: `index_field` plumbing, `_field_depth` / `_flatten_with_coords` / `_walk_value` helpers, optimized `_prepare_batch_db` with batch-level aggregation, `len(self) = len(self.keys)`, new `self._coords` / `self._sample_ids` |
| `fiftyone/core/collections.py` | `to_torch()` accepts `index_field` |
| `tests/unittests/torch_dataset_tests.py` | 256 lines of new tests covering: frame-keyed, label-keyed, depth-mismatched fields, multi-level broadcasts, edge cases |

## Test plan

- [x] Existing unit tests (`tests/unittests/torch_dataset_tests.py`) — all pass
- [x] New tests for arbitrary `index_field` — pass
- [x] End-to-end video inference benchmark on cloud-backed data (see Teams PR companion):
  - DAVIS (1902 frame docs / 30 videos): 130+ fps streaming cold-cache; 15–17× over `_apply_image_model_to_frames_batch`
  - qs_video: 5–7× speedup
  - 100% label parity vs baseline

## Companion PR

- Teams: voxel51/fiftyone-teams `fotd-arbitrary-index-teams` (cherry-pick of this commit + teams-only `required_media_fields` / `_apply_media_cache` integration preserved)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * PyTorch export now accepts a configurable index field, enabling rows to be generated from nested/list fields (e.g., per-detection or per-frame). Dataset length and row keys reflect the flattened indexed rows, and mapped fields broadcast consistently across indexed dimensions.

* **Tests**
  * Added unit tests covering per-detection/per-frame indexing, sibling-branch broadcasting, vectorized vs non-vectorized parity, list-dimension handling, error cases for invalid index paths, and skip-failures behavior.

[Review Change Stack](https://app.coderabbit.ai/change-stack/voxel51/fiftyone/pull/7651?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: fiftyone/core/collections.py [hunk 1]
```diff
@@ -10361,6 +10361,7 @@ def export(
     def to_torch(
         self,
         get_item,
+        index_field="id",
```
Comment (kaixi-wang) [inline fiftyone/core/collections.py:10421]: nit: maybe use key_field? index might give the wrong impression if the field isn't in the index (althouth it prob should if they are using it as a key field)

Comment (brimoor) [reply fiftyone/core/collections.py:10421]: In the context of torch datasets, "index" is the standard nomenclature for what this does (unique identifier for each element of the dataset), so `index_field` is natural here. 

Either way, the description is clear so I think users will grok it

File: tests/unittests/torch_dataset_tests.py [hunk 2]
```diff
@@ -57,14 +57,81 @@ def __exit__(self, *args):
...
+
+class ShortLivedVideoDataset:
+    """In-memory dataset with video samples containing frames."""
+
+    def __init__(self, n_frames_per_sample=(2, 3), **kwargs):
+        self._dataset = fo.Dataset(**kwargs)
+
+        samples = []
+        for i, n in enumerate(n_frames_per_sample):
+            sample = fo.Sample(filepath=f"video{i}.mp4", media_type="video")
```
Comment (brimoor) [inline tests/unittests/torch_dataset_tests.py]: nit: remove explicit `media_type` declaration. It's unnecessary

File: tests/unittests/torch_dataset_tests.py [hunk 3]
```diff
@@ -57,14 +57,81 @@ def __exit__(self, *args):
...
+                            0.2,
+                        ],
+                    )
+                    for j in range(n)
+                ]
+            )
+            samples.append(sample)
+
+        self._dataset.add_samples(samples)
+        self._dataset.persistent = True
```
Comment (brimoor) [inline tests/unittests/torch_dataset_tests.py]: The context manager is unnecessary here. Just create a non-persistent dataset and rely on the fact that FO itself will cleanup all non-persistent datasets when the Python process exits.

All other unit tests that create datasets are successfully using the pattern I've described.

File: fiftyone/utils/torch.py [hunk 4]
```diff
@@ -1847,20 +1848,134 @@ def _is_string_array(targets):
...
+    -  ``"ground_truth.detections.label"`` -> ``("", "ground_truth.detections")``
+
+    The names let us identify the **shared** list prefix between two paths
+    via :func:`_shared_list_prefix_len`; cf. ``index_field`` semantics in
+    :class:`FiftyOneTorchDataset`.
+    """
+    dims = [""]
+    keys = path.split(".")
+
+    if keys[0] == "frames" and samples._has_frame_fields():
```
Comment (manushreegangwar) [inline fiftyone/utils/torch.py]: Could you replace "frames" with `samples._FRAMES_PREFIX`?

File: fiftyone/utils/torch.py [hunk 5]
```diff
@@ -1847,20 +1848,134 @@ def _is_string_array(targets):
...
+    """Indexes into ``nested`` using the first ``depth`` elements of ``coord``.
+
+    Parent fields stop short of the row's coord length; child fields walk the
+    full depth and return whatever subtree remains.
+    """
+    value = nested
+    for i in range(depth):
+        if value is None:
+            return None
+        value = value[coord[i]]
```
Comment (manushreegangwar) [inline fiftyone/utils/torch.py]: Would be good to add a check on `value` type to be list

**Review by manushreegangwar — COMMENTED** [review]
Minor comments. LGTM otherwise!

File: fiftyone/utils/torch.py (hunk 4 above)
Comment (jacobsela) [reply fiftyone/utils/torch.py]: I'm not sure this is a good fit. The check that I wrote here appears elsewhere in the code, e.g. [here](https://github.com/voxel51/fiftyone/blob/e2f0bc65ab965e0f99d413d51576b5b7d5ad94e9/fiftyone/core/collections.py#L1550). I think it's best we follow precedent.

File: fiftyone/utils/torch.py (hunk 5 above)
Comment (jacobsela) [reply fiftyone/utils/torch.py]: Good catch, adding a check and a test.

**Review by manushreegangwar — APPROVED** [review]
🚀

File: tests/unittests/torch_dataset_tests.py (hunk 3 above)
Comment (jacobsela) [reply tests/unittests/torch_dataset_tests.py]: removed

File: tests/unittests/torch_dataset_tests.py (hunk 2 above)
Comment (jacobsela) [reply tests/unittests/torch_dataset_tests.py]: removed

File: fiftyone/utils/torch.py [hunk 6]
```diff
@@ -1847,20 +1848,143 @@ def _is_string_array(targets):
         return False
 
 
+def _list_dims(samples, path):
```
Comment (kaixi-wang) [inline fiftyone/utils/torch.py:1866]: I feel like this parsing is handled elsewhere already? Otherwise, it feels fragile to live here. For example, patches isn't included but essentially the same idea...

File: fiftyone/utils/torch.py [hunk 7]
```diff
@@ -1901,16 +2031,42 @@ def __init__(
...
+            or get_local_rank(local_process_group) == 0
         )
+        if on_master:
+            nested = samples.values(index_field)
+            keys = []
+            coords = []
+            for value, coord in _flatten_with_coords(nested):
+                keys.append(value)
+                coords.append(coord)
+            sample_ids = keys if index_field == "id" else samples.values("id")
```
Comment (kaixi-wang) [inline fiftyone/utils/torch.py]: I think you can get rid of all the helpers and just use a variant of `samples.values([id, field], unwind=True)`

**Review by kaixi-wang — CHANGES_REQUESTED** [review]
i think alot of what you are trying to do is already handled with unwind=True

File: fiftyone/utils/torch.py (hunk 7 above)
Comment (jacobsela) [reply fiftyone/utils/torch.py]: Yes, but we would rather not use `unwind=True` directly to avoid eagerly materializing the entire flattened table to memory. This could potentially be a very large blow up in memory footprint with `vectorize=True` (e.g. repeat any sample level information per frame in a video dataset).

Comment (jacobsela) [reply fiftyone/utils/torch.py]: if you have a clever way to get the best of both worlds here tell me I'll happily change it.

File: fiftyone/utils/torch.py (hunk 6 above)
Comment (jacobsela) [reply fiftyone/utils/torch.py:1866]: good catch, fixing

File: fiftyone/utils/torch.py (hunk 7 above)
Comment (jacobsela) [reply fiftyone/utils/torch.py]: @kaixi-wang restructured things a little, moved the helper into batch construction, and made all `values` calls happen at once to guarantee order.

**Review by kaixi-wang — APPROVED** [review]
